Issue #6 Redis clients not disconnecting

manjiri-g opened the pull request:
Solution: UNSUBSCRIBE and disconnect from Redis on message generator.close() event

Problem:
1. HTTP client to HTTP server: GET /stream
2. HTTP server to Redis server: SUBSCRIBE
3. HTTP client sends TCP FIN
4. Redis client to Redis server: PUBLISH <message1>
5. Flask SSE pubsub.listen yields message1
6. HTTP Response cannot be sent to HTTP client
7. Flask SSE generator.close() gets called

Note: HTTP server will continue to send Response while the front TCP connection
is not CLOSED. If no message gets published after HTTP client has disconnected,
the connection will remain in CLOSE_WAIT state, and therefore Redis client will
remain connected and subscribed. A periodic health check (to be ignored by
HTTP client) gets around this issue.

Commit:
Issue #6 Redis clients not disconnecting

Solution: UNSUBSCRIBE and disconnect from Redis on message generator.close() event

Problem:
1. HTTP client to HTTP server: GET /stream
2. HTTP server to Redis server: SUBSCRIBE
3. HTTP client sends TCP FIN
4. Redis client to Redis server: PUBLISH <message1>
5. Flask SSE pubsub.listen yields message1
6. HTTP Response cannot be sent to HTTP client
7. Flask SSE generator.close() gets called

Note: HTTP server will continue to send Response while the front TCP connection
is not CLOSED. If no message gets published after HTTP client has disconnected,
the connection will remain in CLOSE_WAIT state, and therefore Redis client will
remain connected and subscribed. A periodic health check (to be ignored by
HTTP client) gets around this issue.

Commit:
Add test for redis.exceptions.ConnectionError code coverage

singingwolfboy commented on flask_sse.py:
```diff
@@ -135,10 +136,23 @@ def messages(self, channel='sse'):
+            except redis_exceptions.ConnectionError:
+                pass
+
+    def messages_loop(self, channel, pubsub):
```
Why are you breaking this out into a separate method? I think it's fine to leave it inline.

manjiri-g replied:
I agree. The separate method is for something I am still working on..... Will change.

singingwolfboy commented on flask_sse.py:
```diff
@@ -135,10 +136,23 @@ def messages(self, channel='sse'):
+        finally:
+            try:
+                pubsub.unsubscribe(channel)
+            except redis_exceptions.ConnectionError:
```
Interesting... when would we have a `ConnectionError` here?

manjiri-g replied:
If pubsub.listen() encounters a connection error e.g. if you stop the Redis server, the loop terminates due to an exception. We get "exception while handling exception" in pubsub.unsubscribe() in this scenario.

singingwolfboy commented on tests/test_blueprint.py:
```diff
@@ -95,13 +96,36 @@ def test_messages_channel(bp, app, mockredis):
+    pubsub.unsubscribe.assert_called_with('whee')
+
+
+def test_messages_close(bp, app, mockredis):
```
Thanks for updating the existing tests, and adding this one!

manjiri-g replied:
You're welcome! Was easy to do.

singingwolfboy reviewed (changes requested):
Thanks a lot for sending this pull request! This actually makes a lot of sense, and seems like a good fix for #6. But I think the code can be simplified. What if we just do this?

```python
    def messages(self, channel='sse'):
        """
        A generator of :class:`~flask_sse.Message` objects from the given channel.
        """
        pubsub = self.redis.pubsub()
        pubsub.subscribe(channel)
        try:
            for pubsub_message in pubsub.listen():
                if pubsub_message['type'] == 'message':
                    msg_dict = json.loads(pubsub_message['data'])
                    yield Message(**msg_dict)
        finally:
            pubsub.unsubscribe(channel)
```

manjiri-g commented:
> Thanks a lot for sending this pull request! This actually makes a lot of sense, and seems like a good fix for #6. But I think the code can be simplified. What if we just do this?
> 
> ```python
>     def messages(self, channel='sse'):
>         """
>         A generator of :class:`~flask_sse.Message` objects from the given channel.
>         """
>         pubsub = self.redis.pubsub()
>         pubsub.subscribe(channel)
>         try:
>             for pubsub_message in pubsub.listen():
>                 if pubsub_message['type'] == 'message':
>                     msg_dict = json.loads(pubsub_message['data'])
>                     yield Message(**msg_dict)
>         finally:
>             pubsub.unsubscribe(channel)
> ```

Yes, of course. This was simply to allow subclassing ServerSentEventsBlueprint without having to deal with class:Message. It came about because I am subclassing ServerSentEventsBlueprint to tinker with messages_loop. The idea is to replace pubsub.listen() with pubsub.get_message() and do something useful on timeout. 

The disconnect on the HTTP side is discovered only when a message is published. The app may not publish one for a long time.

Still, the simplification you have suggested makes good sense. I will make that change.

singingwolfboy commented:
> This was simply to allow subclassing ServerSentEventsBlueprint without having to deal with class:Message.

What's wrong with the `Message` class? I made it just as a way of representing a server-sent event, and it's basically a dictionary that knows how to serialize itself according to the SSE spec. If you want to turn it back into a regular dictionary, you can always call `.to_dict()` on it, or reference the `__dict__` property.

manjiri-g commented:
> > This was simply to allow subclassing ServerSentEventsBlueprint without having to deal with class:Message.
> 
> What's wrong with the `Message` class? I made it just as a way of representing a server-sent event, and it's basically a dictionary that knows how to serialize itself according to the SSE spec. If you want to turn it back into a regular dictionary, you can always call `.to_dict()` on it, or reference the `__dict__` property.

When I said "not deal with", I meant "hide the details". In other words, user would have had a way to override messages_loop() as a pre-Message processing step. Sorry for the confusion. (I am going to simplify as per your suggestion.)

Commit:
Issue #6 Redis clients not disconnecting

Simplify code as per review feedback:
https://github.com/singingwolfboy/flask-sse/pull/24#pullrequestreview-448884597

manjiri-g commented:
> Thanks a lot for sending this pull request! This actually makes a lot of sense, and seems like a good fix for #6. But I think the code can be simplified. What if we just do this?
> 
> ```python
>     def messages(self, channel='sse'):
>         """
>         A generator of :class:`~flask_sse.Message` objects from the given channel.
>         """
>         pubsub = self.redis.pubsub()
>         pubsub.subscribe(channel)
>         try:
>             for pubsub_message in pubsub.listen():
>                 if pubsub_message['type'] == 'message':
>                     msg_dict = json.loads(pubsub_message['data'])
>                     yield Message(**msg_dict)
>         finally:
>             pubsub.unsubscribe(channel)
> ```

Done!

singingwolfboy commented on tests/test_blueprint.py:
```diff
@@ -95,13 +96,36 @@ def test_messages_channel(bp, app, mockredis):
             "data": '{"data": "whee", "id": "abc"}',
         }
     ]
+    pubsub.unsubscribe.side_effect = redis.exceptions.ConnectionError()
```
You still have this `ConnectionError` in the tests, but it looks like it got dropped from the code being tested. Also, it looks like GitHub had a temporary outage, and the tests didn't get run. Can you push another commit to restore the `ConnectionError` handling, and make sure the tests run and pass?

manjiri-g replied:
Good catch! I added back this code that got accidentally removed.

singingwolfboy commented on flask_sse.py:
```diff
@@ -4,6 +4,7 @@
 from collections import OrderedDict
 from flask import Blueprint, request, current_app, json, stream_with_context
 from redis import StrictRedis
+from redis import exceptions as redis_exceptions
```
You can probably just do:
```suggestion
from redis.exceptions import ConnectionError
```

Commit:
Issue #6 Redis clients not disconnecting
Handle redis.exceptions.ConnectionError (accidentally deleted in previous commit)

manjiri-g committed:
Update flask_sse.py

Co-authored-by: David Baumgold <david@davidbaumgold.com>

Commit:
Merge branch 'generator-close' of https://github.com/manjiri-g/flask-sse into generator-close

Commit:
Change redis_exceptions.ConnectionError to ConnectionError

frnsys commented:
Would it be possible to merge this in soon?

I also found it helpful to have `pubsub.close()` after `pubsub.unsubscribe(channel)`

singingwolfboy commented:
Thanks for reminding me about this pull request! Everything looks good, so I've merged it and published Flask-SSE version 1.0! 🎉
